rfcs/0002-ethereum-tracing-cache.md

```diff
@@ -0,0 +1,61 @@
+# RFC-0002: Ethereum Tracing Cache
+
+<dl>
+  <dt>Author</dt>
+  <dd>Zac Burns</dd>
+
+  <dt>RFC pull request</dt>
+  <dd><a href="https://github.com/graphprotocol/rfcs/pull/4">https://github.com/graphprotocol/rfcs/pull/4</a></dd>
+
+  <dt>Obsoletes (if applicable)</dt>
+  <dd>None</dd>
+
+  <dt>Date of submission</dt>
+  <dd>2019-12-13</dd>
+
+  <dt>Date of approval</dt>
+  <dd>YYYY-MM-DD</dd>
+
+  <dt>Approved by</dt>
+  <dd>First Person, Second Person</dd>
+</dl>
+
+## Summary
+
+This RFC proposes the creation of a local ethereum tracing cache to speed up indexing of subgraphs which use block and/or call handlers.
+
+## Motivation
+
+When indexing a subgraph that uses block and/or call handlers, it is necessary to extract calls from the trace of each block that a Graph Node indexes. It is expensive to acquire and process traces from ethereum nodes in both money and time.
+
+When developing a subgraph it is common to make changes and deploy those changes to a production Graph Node for testing. Each time a change is deployed, the Graph Node must re-sync the subgraph using the same traces that were used for the previous sync of the subgraph. The cost of acquiring the traces each time a change is deployed impacts a subgraph developer's ability to iterate and test quickly.
+
+## Urgency
+
+None
+
+## Terminology
+
+Ethereum cache: The new API proposed here.
+
+## Detailed Design
+
+There is an existing `EthereumCallCache` for caching `eth_call` built into Graph Node today. This cache will be extended to support traces, and renamed to `EthereumCache`.
+
+## Compatibility
+
+This change is backwards compatible. Existing code can continue to use the parity tracing API. Because the cache is local, each indexing node may delete the cache should the format or implementation of caching change. In this case of invalidated cache the code will fall back to existing methods for retrieving a trace and repopulating the cache.
+
+## Drawbacks and Risks
+
+Subgraphs which are not being actively developed will incur the overhead for storing traces, but will not ever reap the benefits of ever reading them back from the cache.
```

Comment:
This is a great point. We could eventually remove data from the cache (actually, both caches — the `eth_call` and the `trace_filter` one) if the contracts the data is for are no longer used by any subgraphs. At the same time we don't want to do this too eagerly because someone might deploy a subgraph again that will use the cached data. Then again, all it does is affect performance, so cleanups will not do much harm.

Let's keep it in the back of our minds for now. Having this note here is good.